### PR TITLE
Working MVP of leaderboard toggle

### PR DESCRIPTION
**Pull request by drepram**
> Is there any issue / PR I can reference in the title?

I've added the toggle that would trigger the leader-board between the latest or best submission, I've added the component in the UI and the API call has been written into the code.

The problem lies with the Unauthorized call, here's what I've been getting.

Error 401

What it normally looks like:
Normal UI

Please take a look, feedbacks are wanted.

#### Changes proposed in this pull request:

- Toggle to sort leaderboard by latest or best submission branch

- Link to live demo: http://pr-252-evalai.surge.sh 

- Normal UI

### CODE REVIEW

**Comment by Sanji515** [comment]
> The problem lies with the Unauthorized call

Hey @drepram You are not logged in on the screenshot that you shared, in order to toggle the leaderboard by latest or best submission you need to be logged in as challenge host.

**Comment by pushkalkatara** [comment]
Maybe we can create the arrows near the fields as a filter to re-arrange the submissions. Like clicking on up arrow of the score would give all the submissions from high to low order.

**Comment by drepram** [comment]
> Maybe we can create the arrows near the fields as a filter to re-arrange the submissions. Like clicking on up arrow of the score would give all the submissions from high to low order.

I think we can already do this. Natively from the table component.

Please take a look at the latest commit @Sanji515.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.html [hunk 1]
```diff
@@ -41,6 +41,14 @@
                 <mat-chip-list>
                   <mat-chip>B</mat-chip> - Baseline submission
                 </mat-chip-list>
+                <mat-chip-list class="pull-right" *ngIf="isChallengeHost">
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.html:44]: Hey @drepram, I think you also need to get the boolean value for `isChallengeHost` in the typescript.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.html [hunk 2]
```diff
@@ -41,6 +41,14 @@
...
+                  <div (click)="toggleShowLeaderboardByLatest()" class="btn-switch sort-leaderboard-switch" [ngClass]="{'btn-switch--on':selectedPhaseSplitLeaderboardToggle}"
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.html]: You also need to update this function name here.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.scss [hunk 3]
```diff
@@ -82,3 +82,40 @@
...
+  .btn-switch-circle--on {
+      left: auto;
+      right: 0;
+  }
+}
+
+.sort-leaderboard-switch {
+	vertical-align: middle;
+	display: inline-block;
+}
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.scss]: Please add a new line here.

File: src/app/services/endpoints.service.ts [hunk 4]
```diff
@@ -208,6 +208,14 @@ export class EndpointsService {
     return `${this.jobs}challenge_phase_split/${phaseSplitId}/leaderboard/?page_size=1000`;
   }
 
+  /**
+   * Challenge Leaderboard fetch
```
Comment (Sanji515) [inline src/app/services/endpoints.service.ts]: Can you please change this comment here to `Returns or Updates challenge phase split`?

File: src/app/services/endpoints.service.ts [hunk 5]
```diff
@@ -208,6 +208,14 @@ export class EndpointsService {
...
+   * @param phaseSplitId  phase split id
+   */
+  challengeLatestSubmissionURL(phaseSplitId) {
```
Comment (Sanji515) [inline src/app/services/endpoints.service.ts]: Can we please change this to `particularChallengePhaseSplitUrl` or something like this?

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 6]
```diff
@@ -87,12 +87,18 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
    * Currently selected phase split
    */
   selectedPhaseSplit: any = null;
+  selectedPhaseSplitLeaderboardToggle = false;
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Instead of setting this default value to `false`, you should get default value from the backend (what if challenge host changes the default value of this while creating challenge.)

Even how about if we use `show_leaderboard_by_latest_submission` directly from `selectedPhaseSplit` object defined in the typescript? and for which I think we need to update the serializer.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 7]
```diff
@@ -404,6 +411,33 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
...
+      () => {}
+    );
+  }
+
+  toggleShowLeaderboard() {
+    const API_PATH = this.endpointsService.challengeLatestSubmissionURL(this.selectedPhaseSplit['id']);
+    const SELF = this;
+    SELF.selectedPhaseSplitLeaderboardToggle = !SELF.selectedPhaseSplitLeaderboardToggle;
+    const BODY = JSON.stringify({
+      'show_leaderboard_by_latest_submission': !SELF.selectedPhaseSplitLeaderboardToggle
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Here why you are updating this boolean variable twice?

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 8]
```diff
@@ -404,6 +411,33 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
...
+        SELF.sortLeaderboardTextOption = SELF.selectedPhaseSplitLeaderboardToggle ? 'Sort by best' : 'Sort by latest';
+      },
+      err => {
+        SELF.globalService.handleApiError(err);
+      },
+      () => {}
+    );
+  }
+
+  toggleShowLeaderboard() {
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Can we change this name to `showLeaderboardByLatestOrBest`?

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 6 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: 1) I've set the default value to null. `!null` would evaluates to `true`.

Screen Shot 2019-12-31 at 18 18 11

I can't seem to see any boolean value regarding the default value of leaderboard sorting in the initialized API calls. Can you tell me where I can find the boolean value so when the API response is received, I can quickly reassign the value of `this.selectedPhaseSplitLeaderboardToggle`.

2) There is no `show_leaderboard_by_latest_submission` in any objects received. Can you tell me where I can find it?

And I can't find any serializer in the code.

---

In another note, are these changes necessary?

Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Hey @drepram In order to get `show_leaderboard_by_latest_submission` from the `selectedPhaseSplit` object, you need to add this field in the serializer on EvalAI repo.

Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: @drepram Field already been added [here](https://github.com/Cloud-CV/EvalAI/pull/2568) so now you can get `show_leaderboard_by_latest_submission` field from the object.

Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Screen Shot 2020-01-02 at 23 57 21

Due to the non-existence of the `show_leaderboard_by_latest_submission` field in the `selectedPhaseSplit` object, I've decided to make a condition for it. Let me know if I need to make more changes, I would love to make them.
```javascript
if (this.phaseSplits[this.selectedPhaseSplit['id']].hasOwnProperty('show_leaderboard_by_latest_submission')) {
  BODY = JSON.stringify({
    'show_leaderboard_by_latest_submission': this.phaseSplits[this.selectedPhaseSplit['id']].show_leaderboard_by_latest_submission
  });
} else {
  BODY = JSON.stringify({
    'show_leaderboard_by_latest_submission': SELF.selectedPhaseSplitLeaderboardToggle
  });
}
```

**Comment by drepram** [comment]
I've decided to fix https://github.com/Cloud-CV/EvalAI-ngx/issues/253 simultaneously, may this be turned into a task too?

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 9]
```diff
@@ -87,12 +92,18 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
    * Currently selected phase split
    */
   selectedPhaseSplit: any = null;
+  selectedPhaseSplitLeaderboardToggle: boolean = null;
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: @drepram I think you need to initialize it with the value which is fetched from the back-end and also please change this variable name to `showLeaderboardByLatest`.

**Comment by Sanji515** [comment]
> I've decided to fix #253 simultaneously, may this be turned into a task too?

Please open a separate PR for this.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 6 above)
Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: @drepram Please try rebuilding the django container then you'll be able to get that field from the `selectedPhaseSplit` object.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 9 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Like I said before, what if the variable is not present in the objects retrieved from the back-end. Please read my previous messages on our previous conversation here.

Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: This will not be the case, there will always be this field with the default value as `False`.

Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Changed the name.

**Comment by drepram** [comment]
I still can't retrieve the `how_leaderboard_by_latest_submission` field in the `selectedPhaseSplit` object, I've asked @nikochiko if he was able to retrieve it, and the results are the same. We both can't retrieve it, I think it's best to make a condition for it just to be safe.

> Please open a separate PR for this.

I sure will, after this.

**Comment by Sanji515** [comment]
> I still can't retrieve the how_leaderboard_by_latest_submission field in the selectedPhaseSplit object

Have you re-builded the django container?

**Comment by drepram** [comment]
> > I still can't retrieve the how_leaderboard_by_latest_submission field in the selectedPhaseSplit object
> 
> Have you re-builded the django container?

I have, still nothing. @nikochiko did too and he can't.

Yes, I have run `docker-compose up --build`, what exactly would the problem be if we already have a condition that is written to handle it?

**Comment by Sanji515** [comment]
> Yes, I have run `docker-compose up --build`, what exactly would the problem be if we already have a condition that is written to handle it?

That condition is not required there. I've checked it on my local after re-building, field `show_leaderboard_by_latest_submission` is showing on `selectedPhaseSplit ` object.

Also you need to initialize the variable `showLeaderboardByLatest` with this value when phase gets selected or you'll always show the default value as `False` even if its default value is `True` on the back-end.

**Comment by drepram** [comment]
Sanjeev, I've ran `docker system prune && docker-compose down && docker-compose up --build` more than 5 times and I still don't see the fields you've added, perhaps the fields doesn't exist in dummy challenges?

Come on Sanjeev, please don't make this harder, the necessary changes were made and this new API thing was your idea. I stayed up from 12 am to 5 am just trying to see if I can conform to the changes you've made, but I can't.

So please Sanjeev, is the retrieval of a particular challenge leaderboard sorting choice that necessary? The task was about `adding a toggle to switch between latest or best submission`, it wasn't necessarily about changing the API endpoint behavior, is it?

I'm trying so hard to not be rude and disrespectful and am sorry if this message is interpreted in that manner, but the essential part and arguably the task has been completed, hasn't it?

And the thing about `showLeaderboardByLatest`, the default value is false. Arguably, people always want to see the best one first, rather than the latest ones.

I'm sorry if this comes off as rude.

**Comment by Sanji515** [comment]
Hey @drepram 

> Sanjeev, I've ran docker system prune && docker-compose down && docker-compose up --build more than 5 times and I still don't see the fields you've added, perhaps the fields doesn't exist in dummy challenges?

Okay this is weird. I checked on my local, the field was showing. Okay then leave it with the conditions for now.

> And the thing about showLeaderboardByLatest, the default value is false. Arguably, people always want to see the best one first, rather than the latest ones.

If you will not set the default value to as on the back-end field [here](https://github.com/Cloud-CV/EvalAI/blob/master/apps/challenges/models.py#L330) then it may happen that by default the leaderboard comes as latest entries on top but still the toggle shows text as `sort by latest` due to this condition [here](https://github.com/Cloud-CV/EvalAI/blob/master/apps/jobs/views.py#L568).
That's why I added that field on the serializer so that you can get that field from `selectedPhaseSplit` object when the phase split get selected [here](https://github.com/Cloud-CV/EvalAI-ngx/blob/master/src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts#L235) and set its default value.

> The task was about adding a toggle to switch between latest or best submission, it wasn't necessarily about changing the API endpoint behavior, is it?

The task was to add front-end and connect it to the back-end to be able to toggle b/w the latest or best submission on the leader-board page. And API doesn't change here, API remains the same, here on [this](https://github.com/Cloud-CV/EvalAI/pull/2568) PR it just add the field into the serializer so that you can set the default value when the phase split is selected.

> I'm sorry if this comes off as rude.

No no this wasn't rude at all. Happy Coding:smile: 


Please let me know if I'm not clear.

**Comment by drepram** [comment]
I've made the condition, please take a look.

Thanks!

**Comment by Sanji515** [comment]
Screenshot from 2020-01-05 18-18-44


@drepram please see this case here(^) which I've discussed on the above comment. On the backend my default value for `show_leaderboard_by_latest_submission` is set `True` but on the UI still showing to toggle by `show by latest`.

**Comment by drepram** [comment]
Is that behaviour not correct?

`show_leaderboard_by_latest_submission` is set to `true`, Henceforth the leaderboard will be sorted by latest. It might be worth to note that I swapped the behaviour of the ternary operator from the original EvalAI [code](https://github.com/Cloud-CV/EvalAI/blob/master/frontend/src/js/controllers/challengeCtrl.js#L650) because of my common sense, should I revert this change of mine?

**Comment by Sanji515** [comment]
> Is that behaviour not correct?
> 
> `show_leaderboard_by_latest_submission` is set to `true`, Henceforth the leaderboard will be sorted by latest. It might be worth to note that I swapped the behaviour of the ternary operator from the original EvalAI [code](https://github.com/Cloud-CV/EvalAI/blob/master/frontend/src/js/controllers/challengeCtrl.js#L650) because of my common sense, should I revert this change of mine?

No, it is not correct. As you said if this field default set to `True` then leaderboard sorted as latest then still why on the UI its showing `show by latest` if its already sorted as latest. It should ask the user to `sort by best`.

**Comment by drepram** [comment]
> > Is that behaviour not correct?
> > `show_leaderboard_by_latest_submission` is set to `true`, Henceforth the leaderboard will be sorted by latest. It might be worth to note that I swapped the behaviour of the ternary operator from the original EvalAI [code](https://github.com/Cloud-CV/EvalAI/blob/master/frontend/src/js/controllers/challengeCtrl.js#L650) [](/Cloud-CV/EvalAI/blob/HEAD@{2020-01-05T16:50:08Z}/frontend/src/js/controllers/challengeCtrl.js#L650) because of my common sense, should I revert this change of mine?
> 
> No, it is not correct. As you said if this field default set to `True` then leaderboard sorted as latest then still why on the UI its showing `show by latest` if its already sorted as latest. It should ask the user to `sort by best`.

So if the record on the database wants the leaderboard to be sorted by latest, we should ask the user if they want to sort it by best instead? I don't get it, can you explain?

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 10]
```diff
@@ -245,6 +260,15 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
       SELF.selectedPhaseSplit = phaseSplit;
       if (SELF.selectedPhaseSplit) {
         SELF.fetchLeaderboard(SELF.selectedPhaseSplit['id']);
+        if (SELF.phaseSplits[SELF.selectedPhaseSplit['id']].hasOwnProperty('show_leaderboard_by_latest_submission')) {
+          const toggleByLatest = SELF.phaseSplits[SELF.selectedPhaseSplit['id']].hasOwnProperty('show_leaderboard_by_latest_submission');
+
+          if (toggleByLatest === true) {
+            SELF.showLeaderboardByLatest = true;
+          } else {
+             SELF.showLeaderboardByLatest = false;
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: @drepram Here you should directly set `SELF.showLeaderboardByLatest = SELF.selectedPhaseSplit.show_leaderboard_by_latest_submission`

**Comment by Sanji515** [comment]
> So if the record on the database wants the leaderboard to be sorted by latest, we should ask the user > if they want to sort it by best instead? I don't get it, can you explain?

Yes, even if you look  [here](https://github.com/Cloud-CV/EvalAI/blob/master/frontend/src/js/controllers/challengeCtrl.js#L649) then also if `show_leaderboard_by_latest_submission` field set to `True` then text is equal to `Sort by best`.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 10 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Can I make a ternary operator that would be a fallback if there is no value?

```javascript
SELF.showLeaderboardByLatest = ('selectedPhaseSplit.show_leaderboard_by_latest_submission' in self ? selectedPhaseSplit.show_leaderboard_by_latest_submission : true)
```

^ this is for the true case, for the false case it would be different

**Comment by drepram** [comment]
> > So if the record on the database wants the leaderboard to be sorted by latest, we should ask the user > if they want to sort it by best instead? I don't get it, can you explain?
> 
> Yes, even if you look [here](https://github.com/Cloud-CV/EvalAI/blob/master/frontend/src/js/controllers/challengeCtrl.js#L649) [](/Cloud-CV/EvalAI/blob/HEAD@{2020-01-05T16:57:34Z}/frontend/src/js/controllers/challengeCtrl.js#L649) then also if `show_leaderboard_by_latest_submission` field set to `True` then text is equal to `Sort by best`.

So I need to swap the ternary operator, basically?

**Comment by Sanji515** [comment]
> So I need to swap the ternary operator, basically?

Yes, but [this](https://github.com/Cloud-CV/EvalAI-ngx/pull/252#discussion_r363104718) also.

**Comment by drepram** [comment]
> > So I need to swap the ternary operator, basically?
> 
> Yes, but [this](https://github.com/Cloud-CV/EvalAI-ngx/pull/252#discussion_r363104718) also.

Can I use the ternary operator that I've sent to you on that thread?

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 10 above)
Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: `SELF.showLeaderboardByLatest = SELF.selectedPhaseSplit.show_leaderboard_by_latest_submission` This is more cleaner I think. @drepram

File: src/app/components/auth/auth.component.ts [hunk 11]
```diff
@@ -18,7 +18,7 @@ export class AuthComponent implements OnInit {
    */
   localRouter: any;
 
-  @ViewChild('authContainer') authContainer: ElementRef;
+  @ViewChild('authContainer', {read: null}) authContainer: ElementRef;
```
Comment (Sanji515) [inline src/app/components/auth/auth.component.ts]: Hey @drepram Why we need this change here? If not necessary then can you please remove?

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 12]
```diff
@@ -87,12 +92,18 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
...
 
   /**
    * Sort leaderboard based on this column
    */
   sortColumn = 'rank';
 
+  /**
+   * Text option for leadeboard sort
+   */
+  sortLeaderboardTextOption = 'Sort by latest';
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Can we type cast here just `sortLeaderboardTextOption: string;`? As we gonna initliaze when we select phase.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 13]
```diff
@@ -245,6 +260,9 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
       SELF.selectedPhaseSplit = phaseSplit;
       if (SELF.selectedPhaseSplit) {
         SELF.fetchLeaderboard(SELF.selectedPhaseSplit['id']);
+        if (SELF.phaseSplits[SELF.selectedPhaseSplit['id']].hasOwnProperty('show_leaderboard_by_latest_submission')) {
+          SELF.showLeaderboardByLatest = SELF.selectedPhaseSplit.show_leaderboard_by_latest_submission;
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Hey @drepram I've tested the field is showing on my local and there's not gonna be any case for this field to be missing so you can remove these if conditions and also you should initialize here the text variable as I've said on the above comment. So total change on this section should be like

`SELF.showLeaderboardByLatest = SELF.selectedPhaseSplit.show_leaderboard_by_latest_submission;`
`SELF.sortLeaderboardTextOption = SELF.showLeaderboardByLatest ? 'Sort by best' : 'Sort by latest';`

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 14]
```diff
@@ -404,6 +422,42 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
       data => {
         SELF.updateLeaderboardResults(data['results'], SELF);
         SELF.startLeaderboard(SELF.selectedPhaseSplit['id']);
+        SELF.sortLeaderboardTextOption = SELF.showLeaderboardByLatest ? 'Sort by best' : 'Sort by latest';
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: How about instead of making any change here on `refreshLeaderboard` function we update `sortLeaderboardTextOption` after success API call of `showLeaderboardByLatestOrBest` function?

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 15]
```diff
@@ -404,6 +422,42 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
...
+      () => {}
+    );
+  }
+
+  showLeaderboardByLatestOrBest() {
+    const API_PATH = this.endpointsService.particularChallengePhaseSplitUrl(this.selectedPhaseSplit['id']);
+    const SELF = this;
+    let BODY;
+
+    SELF.showLeaderboardByLatest = !SELF.showLeaderboardByLatest;
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: It might happen that API gets call return with some error so it we should not update this variable here instead we should update this variable too after the success of API call.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 16]
```diff
@@ -404,6 +422,42 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
...
+
+    if (this.phaseSplits[this.selectedPhaseSplit['id']].hasOwnProperty('show_leaderboard_by_latest_submission')) {
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Same for here as I've explained on the above comment, you can remove these if conditions and can directly assign the `!`(not) of variable like.

`BODY = JSON.stringify({ 'show_leaderboard_by_latest_submission': !SELF.showLeaderboardByLatest });`

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 17]
```diff
@@ -404,6 +422,42 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
...
+    }
+
+    SELF.leaderboard = [];
+    SELF.showLeaderboardUpdate = false;
+    SELF.apiService.patchUrl(
+      API_PATH,
+      BODY
+    ).subscribe(
+      data => {
+        SELF.updateLeaderboardResults(data, SELF);
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: As this function here gets called as you have already called `refreshLeaderboard` so you can remove this function here.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 18]
```diff
@@ -404,6 +422,42 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
...
+        SELF.refreshLeaderboard();
+        SELF.sortLeaderboardTextOption = SELF.showLeaderboardByLatest ? 'Sort by best' : 'Sort by latest';
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts:448]: And please move this variable update thing above the `SELF.refreshLeaderboard();` function.

File: src/app/components/nav/header-static/header-static.component.ts [hunk 19]
```diff
@@ -61,7 +60,7 @@ export class HeaderStaticComponent implements OnInit, OnDestroy {
   routePath = '/dashboard';
 
 
-  @ViewChild('navContainer') navContainer: ElementRef;
+  @ViewChild('navContainer', {read: null}) navContainer: ElementRef;
```
Comment (Sanji515) [inline src/app/components/nav/header-static/header-static.component.ts]: Same here, please remove if not necessary.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 20]
```diff
@@ -404,6 +422,42 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
...
+        'show_leaderboard_by_latest_submission': this.phaseSplits[this.selectedPhaseSplit['id']].show_leaderboard_by_latest_submission
+      });
+    } else {
+      BODY = JSON.stringify({
+        'show_leaderboard_by_latest_submission': SELF.showLeaderboardByLatest
+      });
+    }
+
+    SELF.leaderboard = [];
+    SELF.showLeaderboardUpdate = false;
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Also we don't need this here as it is already gets set on the `refreshLeaderboard` function.

**Review by Sanji515 — COMMENTED** [review]
Hey @drepram If made correctly then after these changes we can merge this PR. Please take a look.

File: src/app/components/nav/header-static/header-static.component.ts (hunk 19 above)
Comment (drepram) [reply src/app/components/nav/header-static/header-static.component.ts]: Screen Shot 2020-01-06 at 22 35 51
Screen Shot 2020-01-06 at 22 36 27

It's necessary, if the opts parameter is not provided, the build will fail.

File: src/app/components/auth/auth.component.ts (hunk 11 above)
Comment (drepram) [reply src/app/components/auth/auth.component.ts]: Screen Shot 2020-01-06 at 22 35 51
Screen Shot 2020-01-06 at 22 36 27

It's necessary, if the opts parameter is not provided, the build will fail.

Comment (drepram) [reply src/app/components/auth/auth.component.ts]: It's a bit tricky with the static field, one version wants it, the other one rejects it. I follow the latter, the one that Travis is using.

File: src/app/components/nav/header-static/header-static.component.ts (hunk 19 above)
Comment (drepram) [reply src/app/components/nav/header-static/header-static.component.ts]: It's a bit tricky with the static field, one version wants it, the other one rejects it. I follow the latter, the one that Travis is using.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 20 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 18 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts:448]: Changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 17 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 16 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 15 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 14 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 13 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 12 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Changed!

File: src/app/components/auth/auth.component.ts (hunk 11 above)
Comment (drepram) [reply src/app/components/auth/auth.component.ts]: Changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 21]
```diff
@@ -411,4 +428,35 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
...
+      });
+    }
+
+    SELF.apiService.patchUrl(
+      API_PATH,
+      BODY
+    ).subscribe(
+      data => {
+        SELF.sortLeaderboardTextOption = SELF.showLeaderboardByLatest ? 'Sort by best' : 'Sort by latest';
+        SELF.showLeaderboardByLatest = !SELF.showLeaderboardByLatest;
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts:447]: You should move this above line 452 cause based on this variable only you are updating `SELF.sortLeaderboardTextOption`.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 16 above)
Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: @drepram You haven't made this change here. Remove this if and else case here and you can just change it to `BODY = JSON.stringify({ 'show_leaderboard_by_latest_submission': !SELF.showLeaderboardByLatest });`

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 22]
```diff
@@ -127,6 +138,7 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
     public: 3,
   };
 
+
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Please remove this change.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 23]
```diff
@@ -87,12 +92,18 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
    * Currently selected phase split
    */
   selectedPhaseSplit: any = null;
+  showLeaderboardByLatest = false;
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts:99]: Can you please give a comment for this variable here?

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 24]
```diff
@@ -256,7 +273,7 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
    * @param self  context value of this
    */
   updateLeaderboardResults(leaderboardApi, self) {
-    const leaderboard = leaderboardApi.slice();
+    const leaderboard = (Array.isArray(leaderboardApi) ? leaderboardApi.slice() : [leaderboardApi]);
```
Comment (Sanji515) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Why is this change required here?

File: src/app/components/auth/auth.component.ts (hunk 11 above)
Comment (Sanji515) [reply src/app/components/auth/auth.component.ts]: I see... if the build is failing without this on the master branch then please open a separate PR for this. Please remove this and the one below as it is not related to this PR.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 24 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Both `fetchLeaderboard()` and `refreshLeaderboard()` calls `updateLeaderboardResults()`, yet the argument `refreshLeaderboard()` gives differs from what `fetchLeaderboard()` gives. The solution to this difference is to wrap an object in array if it's not already an array. Let me know if you want to change this.

Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 23 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts:99]: changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 22 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 21 above)
Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts:447]: changed!

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 24 above)
Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: But they return the same response as they call the same API I think.

Comment (drepram) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Okay, I shall change it then. I don't think it would work though.

Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: yes please

File: src/app/components/nav/header-static/header-static.component.ts [hunk 25]
```diff
@@ -4,7 +4,6 @@ import { AuthService } from '../../../services/auth.service';
 import { RouterModule, Router, ActivatedRoute, NavigationEnd } from '@angular/router';
 import { DOCUMENT } from '@angular/common';
 import {ApiService} from '../../../services/api.service';
-import {el} from '@angular/platform-browser/testing/src/browser_util';
```
Comment (Sanji515) [inline src/app/components/nav/header-static/header-static.component.ts]: Please revert this change here @drepram

Comment (drepram) [reply src/app/components/nav/header-static/header-static.component.ts]: This caused an error since the module was not found. I'll revert it and see if it doesn't cause the build to fail.

**Review by Sanji515 — APPROVED** [review]
Thanks @drepram for working this hard, you did a great job :+1: LGTM! Now it can be merged :tada:

**Comment by drepram** [comment]
Thank you Sanjeev! It's my pleasure.

**Review by RishabhJain2018 — APPROVED** [review]
